---
title: "Class 15 Transcriptomics and the analysis of RNA-Sequence data"
author: "Tom Quach (A15549142)"
date: "11/16/2021"
output:
  pdf_document: default
  html_document: default
---

```{r}
library(BiocManager)
library(DESeq2)
```

Importing, reading, and storing the CSV files 

```{r}
counts <- read.csv("airway_scaledcounts.csv", row.names=1)
metadata <-  read.csv("airway_metadata.csv")
```


```{r}
head(counts)
```

```{r}
head(metadata)
```

```{r}
nrow(counts)
```

> Q1 How many genes are in this data set? 

There are 38694 genes in thsi data set

```{r}
sum(metadata$dex == "control")
```
> Q2. How many ‘control’ cell lines do we have?

We have 4 control cell lines 


First I need to extract all the "control" columns. Then I will take the row-wise mean to get the average count values for all genes in these four experiments 
```{r}
control.inds <- metadata$dex == "control"
control.counts <- counts[, control.inds]
head(control.counts)
```

```{r}
control.mean <- rowMeans(control.counts) 
head(control.mean)
```

> Q3. How would you make the above code in either approach more robust?

In order to make it more robust, we can just use the equation given above so we can get the average the rows even when we add more control or treated group. 

> Q4. Follow the same procedure for the treated samples (i.e. calculate the mean per gene across drug treated samples and assign to a labeled vector called treated.mean)

```{r}
treated.inds <- metadata$dex == "treated"
treated.counts <- counts[, treated.inds]
head(treated.counts)
```

```{r}
treated.mean <- rowMeans(treated.counts) 
head(treated.mean)
```

We will combine our meancount data for bookkeeping purposes.

```{r}
meancounts <- data.frame(control.mean, treated.mean)
```


> Q5 (a). Create a scatter plot showing the mean of the treated samples against the mean of the control samples. 

```{r}
plot(meancounts)
```

> Q5 (b).You could also use the ggplot2 package to make this figure producing the plot below. What geom_?() function would you use for this plot? 

I would use the geom_point() function for this plot 

```{r}
library(ggplot2)
ggplot(data = meancounts, aes(x = control.mean, y = treated.mean)) +
    geom_point()
```


> Q6. Try plotting both axes on a log scale. What is the argument to plot() that allows you to do this?

I would use the log argument to plot in order to 

```{r}
plot(meancounts, log=("xy"))
```

We often use log2 in this field because it has nice mathproperties that make interpretation easier 

```{r}
log2(10/10)
```


```{r}
log2(40/10)
```

```{r}
log2(5/10)
```
We see 0 values for no change and positive values for increases and negative values for decreases. Thsi nice property leads us to work with **log2(fold-chage)** all the time in the genomics and proteomics filled. 

Let's add the **log2(fold-change)** to our "mean-counts" dataframe

```{r}
meancounts$log2fc <- log2(meancounts[,"treated.mean"]/meancounts[,"control.mean"])
head(meancounts)
```


I need to exclude the genes (i.e. rows) with zero counts because we can't say anything about these as we have no data for them 

```{r}
head(meancounts[, 1:2])
```

```{r}
head(meancounts[, 1:2] == 0)
```

```{r}
which( c(F,F,T,T))
```

I can use the **which()** function with the 'arr.ind=TRUE' argument to get the columns and rows where the TRUE values are (i.e the zero counts in our case)

```{r}
zero.vals <- which(meancounts[, 1:2] == 0, arr.ind=TRUE)
head(zero.vals)
```

```{r}
to.rm <- unique(zero.vals[,1])
sort(to.rm)
```

Now remove these from our 'meancounts' dataframe

```{r}
mycounts <- meancounts[-to.rm,]
head(mycounts)
```

```{r}
nrow(mycounts)
```

> Q7. What is the purpose of the arr.ind argument in the which() function call above? Why would we then take the first column of the output and need to call the unique() function?

The purpose of **which()** function with the 'arr.ind=TRUE' argument is to get the columns and rows where the TRUE values are (i.e the zero counts in our case). We use the unique() function to remove any duplicate rows

```{r}
up.ind <- mycounts$log2fc > 2
down.ind <- mycounts$log2fc < (-2)
```

> Q8. Using the up.ind vector above can you determine how many up regulated genes we have at the greater than 2 fc level?

```{r}
sum(up.ind == TRUE)
```
There are 250 genes that are up regulated 

> Q9. Using the down.ind vector above can you determine how many down regulated genes we have at the greater than 2 fc level?

```{r}
sum(down.ind == TRUE)
```
There are 367 genes that are down regulated

> Q10. Do you trust these results? Why or why not?

No, I do not trust these results. We have been doing our analysiz based on fold change which can be large without statistically significant. We have not done anything to determine if the differences we are seeing are significant. Therefore, these data can be very misleading. 


##DESeq2


```{r}
library(DESeq2)
citation("DESeq2")
```

```{r}
dds <- DESeqDataSetFromMatrix(countData=counts, 
                              colData=metadata, 
                              design=~dex)
dds
```

Now we can run DESeq analysis 

```{r}
dds <- DESeq(dds)
```

To get at the results here we use the deseq 'results()' function:

```{r}
res <- results(dds)
res
```

> Q11. Q11. Run the mapIds() function two more times to add the Entrez ID and UniProt accession and GENENAME as new columns called res$entrez, res$uniprot and res$genename

```{r}
library("AnnotationDbi")
library("org.Hs.eg.db")
```


```{r}
res$entrez <- mapIds(org.Hs.eg.db,
                     keys=row.names(res),
                     column="ENTREZID",
                     keytype="ENSEMBL",
                     multiVals="first")

res$uniprot <- mapIds(org.Hs.eg.db,
                     keys=row.names(res),
                     column="UNIPROT",
                     keytype="ENSEMBL",
                     multiVals="first")

res$genename <- mapIds(org.Hs.eg.db,
                     keys=row.names(res),
                     column="GENENAME",
                     keytype="ENSEMBL",
                     multiVals="first")

head(res)
```



## Volcano plots 

```{r}
plot(res$log2FoldChange, res$padj)
```

That is not a useful plot because all the small p-values are hidden at the bottom of the plot and we can't really see them. Log will help.

```{r}
plot(res$log2FoldChange, log(res$padj))
```
We can flip this p-value axis by just putting a minus sign on it. Then we will have the classic volcona plot that the rest of the world uses

```{r}
plot(res$log2FoldChange, -log(res$padj))
```

Finally let's add some color to this plot to draw attention to the genes (i.e. points) we care about - that is those with large fold-change and low pvalues (ie. high -log(pvalues))

```{r}
mycols <- rep("gray", nrow(res))
mycols[ abs(res$log2FoldChange) > 2 ]  <- "red" 

inds <- (res$padj < 0.01) & (abs(res$log2FoldChange) > 2 )
mycols[ inds ] <- "blue"

# Volcano plot with custom colors 
plot( res$log2FoldChange,  -log(res$padj), 
 col=mycols, ylab="-Log(P-value)", xlab="Log2(FoldChange)" )

# Cut-off lines
abline(v=c(-2,2), col="gray", lty=2)
abline(h=-log(0.1), col="gray", lty=2)
```

```{r}
library(pathview)
library(gage)
library(gageData)

data(kegg.sets.hs)
```

```{r}
head(kegg.sets.hs, 2)
```


```{r}
foldchanges = res$log2FoldChange
names(foldchanges) = res$entrez
head(foldchanges)
```

```{r}
keggres = gage(foldchanges, gsets=kegg.sets.hs)
```

```{r}
attributes(keggres)
```
```{r}
head(keggres$less, 3)
```

```{r}
pathview(gene.data=foldchanges, pathway.id="hsa05310")
```
> Q12. Can you do the same procedure as above to plot the pathview figures for the top 2 down-regulated pathways?

```{r}
pathview(gene.data=foldchanges, pathway.id="hsa05332")
```


```{r}
pathview(gene.data=foldchanges, pathway.id="hsa04940")
```

